### ISSUE TITLE
Resource merge error for different versions

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-specification#3361]
Currently the specification defines the following [merge strategy](https://github.com/open-telemetry/opentelemetry-specification/blob/9352c2524ac03b31f4b845323b351d32ab3adb43/specification/resource/sdk.md#merge):

> The resulting resource will have the Schema URL calculated as follows:
> 
> - If the old resource's Schema URL is empty then the resulting resource's Schema
>   URL will be set to the Schema URL of the updating resource,
> - Else if the updating resource's Schema URL is empty then the resulting
>   resource's Schema URL will be set to the Schema URL of the old resource,
> - Else if the Schema URLs of the old and updating resources are the same then
>   that will be the Schema URL of the resulting resource,
> - **Else this is a merging error (this is the case when the Schema URL of the old
>   and updating resources are not empty and are different). The resulting resource is
>   undefined, and its contents are implementation-specific.**

This results in resources that are defined using different versions of the OpenTelemetry semantic conventions to be merged into an undefined resource and results in an errors.

Users of libraries that provide resources (via a detector or directly) are often frustrated when they try to merge these resources together or with their own defined resources because the schema URLs differ and the merge results in an error. This is especially frustrating to them given they do not control the upgrade or downgrade of these resources they get from 3rd party libraries and currently have no way to resolve the error.

Given the telemetry schema exist for OpenTelemetry, this behavior should not be required. It should be possible use the telemetry schema to upgrade/downgrade resources for a user and merge successfully into a single resource they can use at a schema URL they define.

### Existing Implementation

The following is a non-comprehensive survey of what language implementation currently do.

| Language | Error treatment | Returned Resource |
| --- | --- | --- |
| [Go](https://github.com/open-telemetry/opentelemetry-go/blob/65ebe5e50f1c4d5aeb01d9693551bc4aab888149/sdk/resource/resource.go#L160-L193) | Error returned to user | Empty resource returned |
| Java | Warning logged | Merged resource without a schema returned |
| Erlang | Warning logged | Merged resource without a schema returned |
| [Python](https://github.com/open-telemetry/opentelemetry-python/blob/17aa1e862c450d27b3c138859cd08b1ec6deaa7d/opentelemetry-sdk/src/opentelemetry/sdk/resources/__init__.py#L223-L257) | Error logged | Original "old" resource returned |
| .Net | ? | ? |
| C++ | ? | ? |
| ruby | ? | ? |
| [Javascript](https://github.com/open-telemetry/opentelemetry-js/blob/4b04407ba34594d5febc263fe6c79fd4cf7475d5/packages/opentelemetry-resources/src/Resource.ts#L119-L151) | Not supported (?) | Not supported (?) |

@open-telemetry/dotnet-maintainers, @open-telemetry/python-maintainers, @open-telemetry/javascript-maintainers, @open-telemetry/cpp-maintainers @open-telemetry/erlang-maintainers, @open-telemetry/ruby-maintainers: please feel free to edit the table above to correct any information.

cc @trask 

### Proposal

Either:

1. Update the merge strategy to have the SDK upgrade/downgrade a resource in the process of merging
2. Add additional recommendations that SDKs provide upgrade/downgrade functionality for resources

**Comment by dyladan** [comment]
For JS I would say the "Returned Resource" is "Merged resource without a schema returned" even though schema is not supported for resource it is technically true.

**Comment by MrAlias** [comment]
> For JS I would say the "Returned Resource" is "Merged resource without a schema returned" even though schema is not supported for resource it is technically true.

Updated. Thanks

**Comment by pellared** [comment]
I think we now have enough implementation experience to narrow this issue to a safe baseline behavior that does not require schema translation.

The implementation table in the issue is out of date. In particular, Go no longer returns an empty Resource, Java logs at `INFO` rather than warning, Erlang does not report the conflict, and JavaScript and .NET now support Resource Schema URLs.

## Current implementation behavior

| Language | Conflict reporting | Returned Resource |
| --- | --- | --- |
| [Go](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/resource/resource.go) | Returns an error wrapping `ErrSchemaURLConflict` | Returns the normally merged attributes with an empty Schema URL. The caller may use the Resource. [`Detect`](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/resource/auto.go) keeps the URL empty if any conflict occurred in the detector chain. This behavior was introduced in [#4876](https://github.com/open-telemetry/opentelemetry-go/pull/4876). |
| [Java](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java#L148-L175) | Logs at `INFO` | Returns the normally merged attributes with a `null` Schema URL. This is covered by [tests](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java#L214-L228). |
| [Erlang/Elixir](https://github.com/open-telemetry/opentelemetry-erlang/blob/main/apps/opentelemetry/src/otel_resource.erl#L120-L146) | No diagnostic | Returns the normally merged attributes with `schema_url = undefined`, as verified by its [test](https://github.com/open-telemetry/opentelemetry-erlang/blob/main/apps/opentelemetry/test/otel_resource_SUITE.erl#L178-L190). |
| [.NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Resources/Resource.cs) | Emits an `EventSource` warning | Returns the normally merged attributes with a `null` Schema URL. The conflict is sticky, so later merges cannot incorrectly restore a Schema URL. Resource Schema URL support was added in [#7472](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7472). |
| [JavaScript](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-resources/src/ResourceImpl.ts) | Emits a `diag.warn` | Returns the normally merged attributes with an undefined Schema URL. |
| [PHP](https://github.com/open-telemetry/opentelemetry-php/blob/main/src/SDK/Resource/ResourceInfo.php) | Logs a warning | Returns the normally merged attributes with a `null` Schema URL. |
| [Rust](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/resource/mod.rs) | No diagnostic | Returns the normally merged attributes with no Schema URL. |
| [Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/resources/__init__.py) | Logs an error | Returns the original old Resource, discarding the updating Resource. |
| [C++](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/sdk/src/resource/resource.cc) | No diagnostic | Merges the attributes and retains the updating Resource's Schema URL. |
| [Ruby](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/sdk/lib/opentelemetry/sdk/resources/resource.rb) | Not applicable | Resource Schema URL is not supported. |
| [Swift](https://github.com/open-telemetry/opentelemetry-swift-core/blob/main/Sources/OpenTelemetrySdk/Resources/Resource.swift) | Not applicable | Resource Schema URL is not supported. |

The predominant implemented behavior is therefore:

1. merge the attributes according to the normal precedence rules;
2. leave the resulting Resource Schema URL empty; and
3. make the conflict observable using the language's normal error or diagnostic mechanism.

The implementations differ mainly in how they report the conflict and whether a conflict remains sticky across a chain of merges.

## Proposed specification change

I suggest replacing the final Schema URL bullet in [`Resource.Merge`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge) with:

> - Else, the Schema URLs conflict (both are non-empty and different). The SDK
>   SHOULD return a Resource whose attributes are merged according to the rules
>   above and whose Schema URL is empty. The SDK SHOULD make the conflict
>   observable using language-specific error handling. Existing implementations
>   MAY retain their current behavior for backwards compatibility.

I also suggest defining behavior across a sequence of merges:

> When merging a sequence of Resources, once a Schema URL conflict has occurred,
> subsequent merges SHOULD NOT assign a non-empty Schema URL unless the
> implementation can establish that the URL applies to all attributes in the
> resulting Resource.

This avoids the following incorrect result:

```text
Resource A, Schema URL A
  + Resource B, Schema URL B
  = merged Resource, empty Schema URL

  + Resource C, Schema URL C
  = merged Resource incorrectly labeled with Schema URL C
```

An implementation can satisfy this without exposing new public API. For example, .NET tracks an internal conflict bit, while Go's detector aggregation tracks whether any merge returned `ErrSchemaURLConflict` and clears the final Schema URL.

The duplicated requirement in [`Detecting resource information from the environment`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#detecting-resource-information-from-the-environment) should be changed at the same time. I suggest replacing:

> If multiple detectors are combined and the detectors use different non-empty
> Schema URL it MUST be an error since it is impossible to merge such resources.
> The resulting resource is undefined, and its contents are implementation
> specific.

with:

> If multiple detectors are combined and the detectors use different non-empty
> Schema URLs, the SDK SHOULD preserve the attributes merged according to the
> normal Resource merge rules and leave the resulting Resource Schema URL empty.
> The SDK SHOULD make the conflict observable using language-specific error
> handling. Existing implementations MAY retain their current behavior for
> backwards compatibility.

An SDK MAY translate Resources to a common supported schema before merging. Schema translation can remain an optional enhancement rather than a requirement for a successful merge.

## Reasoning

### An empty Schema URL is the safe representation

Attaching either conflicting URL to the merged Resource would claim that the URL describes attributes that may have been produced under a different schema. An empty URL makes no such incorrect claim.

This is also the direction adopted for Entities. Accepted [OTEP 264](https://github.com/open-telemetry/oteps/pull/264) explicitly says that the rules around merging Resources across Schema URLs will be dropped and that entity-specific Schema URLs will replace the Resource-level URL where possible.

[#4266](https://github.com/open-telemetry/opentelemetry-specification/issues/4266) reinforces the broader motivation that composing detectors should preserve independently useful data.

### A Schema URL conflict should not discard otherwise useful Resource data

Resource detectors are intentionally extensible and may be maintained and upgraded independently. A service detector and a cloud detector using different semantic-convention versions can still contribute useful, non-conflicting attributes.

Treating this as an undefined merge has repeatedly caused routine dependency updates to fail at application initialization. Examples include [Go #2341](https://github.com/open-telemetry/opentelemetry-go/issues/2341), [Go #3769](https://github.com/open-telemetry/opentelemetry-go/issues/3769), and [Go contrib #3657](https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3657).

Returning the merged, schemaless Resource follows OpenTelemetry's resilience principles: preserve usable telemetry context while making the degraded condition observable.

### Error reporting must remain language-neutral

Requiring a returned error would not be implementable without API changes in languages whose `merge` method returns only a Resource. Existing SDKs use returned errors, logs at different levels, `EventSource`, or no diagnostic.

The specification should define the desired Resource result and use the existing language-specific error-handling principles for observability, rather than requiring one particular error channel.

### The compatibility allowance is necessary

The Resource SDK specification is Stable, and Python and C++ have established materially different behavior. A `SHOULD` describes the prferred interoperable result for new work, while the explicit `MAY` allows those implementations to avoid breaking existing users.

### Schema translation is complementary, not a baseline merge dependency

Translation can produce a better result when an SDK has the required schema definitions and transformation support. However, baseline Resource merging must also work without network access, a schema registry, or knowledge of the schema contents. The schemaless merged Resource is a deterministic fallback when translation is unavailable.

In short, I propose resolving this issue by specifying **merged attributes plus an empty Resource Schema URL** as the preferred recovery behavior, keeping diagnostics language-specific, and explicitly allowing existing implementationsto retain their behavior for backwards compatibility.

CC @open-telemetry/specs-entities-approvers

**Comment by pellared** [comment]
@open-telemetry/cpp-maintainers @open-telemetry/python-maintainers, PTAL. Maybe you prefer changing the behavior instead of adding `Existing implementations MAY retain their current behavior for backwards compatibility` to the specification?

**Comment by jsuereth** [comment]
> In short, I propose resolving this issue by specifying merged attributes plus an empty Resource Schema URL as the preferred recovery behavior, keeping diagnostics language-specific, and explicitly allowing existing implementations to retain their behavior for backwards compatibility.

IIUC - basically anytime we get into an ambiguous merge situation with resource, we throw away SchemaURL and merge. I think this is the behavior we *need* today, and is basically what the Entities Merge algorithm specifies.

I'd be happy to update the stable portion of the specification here as a "breaking bugfix". 

However - an important data point - we have specification wording where the new Entity merge algorithm *is* meant to replace the existing behaivor and provide a fix -

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/data-model.md#merging-resources

> The current SDK specification outlines a [merge algorithm](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge). This specification updates the algorithm to be compliant with entities. This section will replace that section upon stabilization of entities. SDKs SHOULD NOT update their merge algorithm until full Entity SDK support is provided.

The plan on Entities SIG was to fix this with the entities-aware resource detectors - and "change" / "break" the poor behavior we had before with an algorithm we can stand behind.  I think the specified merge logic *will* have the intended behavior you specify here for all the same reasons.

So the question is - did you want to make a spec change NOW ahead of entities? If so - we need to go update all the pieces of the spec that were NOT planning to do so until entities land.

**Comment by pellared** [comment]
> I'd be happy to update the stable portion of the specification here as a "breaking bugfix".

Yes, I would like to make this specification change now, independently of Entities. I think doing this before Entities has several advantages:

1. Waiting for Entities does not eliminate the compatibility impact. If the Entity algorithm will eventually replace the existing Stable merge algorithm, it merely postpones the same change and bundles it with the larger Entity migration.
2. The fix is independently useful today. Direct Resource merges and existing Resource detectors can preserve useful attributes after a Schema URL conflict without requiring Entity support.
3. A focused change now gives SDK maintainers and users an opportunity to validate the behavior and report compatibility problems before the Entity merge algorithm depends on it.
4. As the implementation survey above shows, merging the attributes and clearing the conflicting Schema URL is already the predominant SDK behavior. The specification would mostly be standardizing proven behavior rather than introducing a new algorithm without implementation experience.

So yes, my preference is a focused fix now, followed by having the Entity merge algorithm build on that corrected baseline.

Just to clarify, by an ambiguous merge here, I mean a conflict between two distinct non-empty Resource Schema URLs, including preserving that conflict across subsequent merges.
I am not proposing to change the existing `A + empty -> A` behavior as part of this issue. A change to `A + empty -> empty (merge conflict)` should be better scoped as a separate (follow-up) issue (I think you wanted this as well if I understood correctly your other comment).

**Comment by jsuereth** [comment]
> I am not proposing to change the existing A + empty -> A behavior as part of this issue. A change to A + empty -> empty (merge conflict) should be better scoped as a separate (follow-up) issue (I think you wanted this as well if I understood correctly your other comment).

Yes - I think this is the correct behavior, particularly once entities are in play.  I also think this matches the principles we both agree to (see below).

Let's start with things we agree to and see where we disagree.

Here's a set of points I *think* we agree on:

- We should treat "empty" schema_url as *unknown* schema, not "i don't care and I'm safe to merge".
- We should *allow* resource merges across schema differences, generally, as practice has shown the existing algorithm is too rigid and lacks the features to work in our current ecosystem.
- `schema_url` when it exists, should accurately describe a Resource and be safe to use for validation / conformance tests in weaver.

here's a set of points I'd like to see if we can agree on:

- We SHOULD NOT block entities work here to stabilize a new merge algorithm to then immediately try to stabilize a second entities merge algorithm. I think this is a mistake - we already have had an entity merge algorithm proposed for QUITE some time.  I don't think the incremental really gains us a lot and looses us a lot of momentum.
- We SHOULD fix the current merge specification so SDKs can, in stable locations, remove poor behavior for resource merge and be spec compliant.

If we agree to both of these I think the next path forward is this:

- The first two set of points *imply* that `A + empty -> empty` is the right behavior going forward.  Doing otherwise means we're treating `empty` differently in the presence of entities vs. not and I think this is a no-go.
- We should issue a PR to "fix" the merge algorithm.  I think this will actually be the entity algorithm defined in the Resource model *but* specfiied in absence of entities.   This *SHOULD* immediately fix this bug *and* allow entities to progress in tandem with fixing this issue.

WDYT?

**Comment by pellared** [comment]
> We should treat "empty" schema_url as unknown schema, not "i don't care and I'm safe to merge".

I would be very cautious here. I find this as a potential breaking change. Has there been any work to check how and where schemaless resources are currently used? For instance, this would change the behavior of our Splunk distribution of OTel Go and it would start emitting empty schema URL which not desired: https://github.com/signalfx/splunk-otel-go/blob/aac87bdf0c46a178a5ca4c601afcfdb2c655b93b/distro/otel.go#L147-L157. Therefore, from the perspective of our distribution this would be a breaking change.

>  We should allow resource merges across schema differences,

Agreed

> `schema_url` when it exists, should accurately describe a Resource and be safe to use for validation / conformance tests in weaver.

I understand this desire, but I think there can be other ways to achieve it if community would prefer to have the "empty" schema_url also as "i don't care and I'm safe to merge". For instance, we could introduce a new `strict_schema_url`  and weaver could have a strict and lenient resource validation. PS. I have not checked if there is any prior art. I wouldn't be surprised if this was already discussed. 

> We SHOULD NOT block entities work here to stabilize a new merge algorithm to then immediately try to stabilize a second entities merge algorithm. 

Agreed.

> We SHOULD fix the current merge specification so SDKs can, in stable locations, remove poor behavior for resource merge and be spec compliant.

Agreed.

**Comment by dyladan** [comment]
We dicussed this in the entity SIG yesterday. Given that the SDKs have inconsistent behavior with each other, with varying levels of spec compliance and willingness to change, it might be best to declare bankruptcy on resource-level schema in favor of entity-level schema.

concretely:
- Leave existing SDK implementations of resource schema handling as they are. If SDK maintainers want to update their SDKs to match spec, great.
- Any merge where schemas don't match will drop the schema. It is impossible to make assertions about what the schema should be in such a state.
- Each Entity has its own schema URL which may or may not match the schema of other Entities in the same resource. Each is treated in isolation from the others.
- The resource-level schema is dropped if any entity is merged which has a schema url that doesn't match the preexisting resource-level schema.
- Weaver will only validate entity schemas

**Comment by dyladan** [comment]
> > We should treat "empty" schema_url as unknown schema, not "i don't care and I'm safe to merge".
> 
> I would be very cautious here. I find this as a potential breaking change. Has there been any work to check how and where schemaless resources are currently used? For instance, this would change the behavior of our Splunk distribution of OTel Go and it would start emitting empty schema URL which not desired: https://github.com/signalfx/splunk-otel-go/blob/aac87bdf0c46a178a5ca4c601afcfdb2c655b93b/distro/otel.go#L147-L157. Therefore, from the perspective of our distribution this would be a breaking change.

I didn't see this part the first time. I believe the "breakingness" of this change would apply also to any `SpanProcessor` that wants to add resource attributes without schema. I believe we're going to talk about this in the spec meeting today.
